File: formal/fstar/experimental_ocaml_glue/hacl_stubs.c
/* hacl_stubs.c — OCaml <-> C FFI for the vendored HACL* Ed25519 + SHA-256
 * primitives (third_party/hacl/).  Realises the crypto assume-vals of
 * VC.DataIntegrity.fst (Verifiable Credentials eddsa-rdfc-2022 pipeline).
 *
 * Crypto sourcing policy: skills/crypto-policy/SKILL.md — "Dont roll our
 * own crypto!", "pursue HACL*".  This file contains NO cryptographic
 * logic: every primitive (Ed25519 sign/verify/keygen, SHA-256) is a call
 * into the vendored, F-star/Low-star-verified HACL* extracted C.  The only
 * logic here is hex <-> byte marshalling around the OCaml string boundary, the
 * same shape as experimental_ocaml_glue/parquet_zstd_stubs.c.
 *
 * Boundary contract (all hex is lowercase, no 0x prefix):
 *   caml_hacl_sha256(msg_bytes)                -> 64-char hex digest
 *   caml_hacl_ed25519_secret_to_public(sk_hex) -> pk_hex (64 chars) | ""
 *   caml_hacl_ed25519_sign(sk_hex, msg_hex)    -> sig_hex (128 chars) | ""
 *   caml_hacl_ed25519_verify(pk_hex,msg_hex,sig_hex) -> bool
 * An empty-string return signals a malformed-length/odd-hex input (the
 * F* wrapper treats that as failure); it never signals a crypto verdict.
 */
#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/alloc.h>
#include <caml/fail.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>

#include "Hacl_Ed25519.h"
#include "Hacl_Hash_SHA2.h"

static const char HEXD[] = "0123456789abcdef";

static int hex_nibble(char c) {
  if (c >= '0' && c <= '9') return c - '0';
  if (c >= 'a' && c <= 'f') return 10 + (c - 'a');
  if (c >= 'A' && c <= 'F') return 10 + (c - 'A');
  return -1;
}

/* Decode a hex OCaml string into a freshly malloc'd byte buffer.
 * Returns 1 on success (sets *out and *out_len), 0 on odd length or a
 * non-hex digit. Caller frees *out on success. */
static int hex_decode(value v_hex, unsigned char **out, size_t *out_len) {
  const char *hex = String_val(v_hex);
  size_t hex_len = caml_string_length(v_hex);
  if ((hex_len % 2) != 0) return 0;
  size_t n = hex_len / 2;
  unsigned char *buf = (unsigned char *)malloc(n ? n : 1);
  if (buf == NULL) caml_failwith("hacl_stubs: allocation failure");
  for (size_t i = 0; i < n; i++) {
    int hi = hex_nibble(hex[2 * i]);
    int lo = hex_nibble(hex[2 * i + 1]);
    if (hi < 0 || lo < 0) { free(buf); return 0; }
    buf[i] = (unsigned char)((hi << 4) | lo);
  }
  *out = buf;
  *out_len = n;
  return 1;
}

/* Encode raw bytes into a fresh OCaml hex string. */
static value hex_encode(const unsigned char *bytes, size_t len) {
  CAMLparam0();
  CAMLlocal1(v_out);
  v_out = caml_alloc_string(len * 2);
  unsigned char *dst = Bytes_val(v_out);
  for (size_t i = 0; i < len; i++) {
    dst[2 * i]     = (unsigned char)HEXD[(bytes[i] >> 4) & 0xF];
    dst[2 * i + 1] = (unsigned char)HEXD[bytes[i] & 0xF];
  }
  CAMLreturn(v_out);
}

/* SHA-256 of the raw bytes of the OCaml string argument. */
CAMLprim value caml_hacl_sha256(value v_msg) {
  CAMLparam1(v_msg);
  CAMLlocal1(v_out);
  uint8_t digest[32];
  /* HACL takes a non-const uint8_t*; String_val is safe to read from. */
  Hacl_Hash_SHA2_hash_256(digest, (uint8_t *)String_val(v_msg),
                          (uint32_t)caml_string_length(v_msg));
  v_out = hex_encode(digest, 32);
  CAMLreturn(v_out);
}

/* Ed25519 private key (32 bytes, hex) -> public key (32 bytes, hex). */
CAMLprim value caml_hacl_ed25519_secret_to_public(value v_sk_hex) {
  CAMLparam1(v_sk_hex);
  CAMLlocal1(v_out);
  unsigned char *sk = NULL; size_t sk_len = 0;
  if (!hex_decode(v_sk_hex, &sk, &sk_len) || sk_len != 32) {
    if (sk) free(sk);
    CAMLreturn(caml_copy_string(""));
  }
  uint8_t pk[32];
  Hacl_Ed25519_secret_to_public(pk, sk);
  free(sk);
  v_out = hex_encode(pk, 32);
  CAMLreturn(v_out);
}

/* Ed25519 sign: (sk_hex[32B], msg_hex) -> signature hex (64 bytes). */
CAMLprim value caml_hacl_ed25519_sign(value v_sk_hex, value v_msg_hex) {
  CAMLparam2(v_sk_hex, v_msg_hex);
  CAMLlocal1(v_out);
  unsigned char *sk = NULL, *msg = NULL; size_t sk_len = 0, msg_len = 0;
  if (!hex_decode(v_sk_hex, &sk, &sk_len) || sk_len != 32) {
    if (sk) free(sk);
    CAMLreturn(caml_copy_string(""));
  }
  if (!hex_decode(v_msg_hex, &msg, &msg_len)) {
    free(sk);
    CAMLreturn(caml_copy_string(""));
  }
  uint8_t sig[64];
  Hacl_Ed25519_sign(sig, sk, (uint32_t)msg_len, msg);
  free(sk); free(msg);
  v_out = hex_encode(sig, 64);
  CAMLreturn(v_out);
}

/* Ed25519 verify: (pk_hex[32B], msg_hex, sig_hex[64B]) -> bool. */
CAMLprim value caml_hacl_ed25519_verify(value v_pk_hex, value v_msg_hex,
                                        value v_sig_hex) {
  CAMLparam3(v_pk_hex, v_msg_hex, v_sig_hex);
  unsigned char *pk = NULL, *msg = NULL, *sig = NULL;
  size_t pk_len = 0, msg_len = 0, sig_len = 0;
  int ok = 0;
  if (hex_decode(v_pk_hex, &pk, &pk_len) && pk_len == 32 &&
      hex_decode(v_msg_hex, &msg, &msg_len) &&
      hex_decode(v_sig_hex, &sig, &sig_len) && sig_len == 64) {
    ok = Hacl_Ed25519_verify(pk, (uint32_t)msg_len, msg, sig) ? 1 : 0;
  }
  if (pk) free(pk);
  if (msg) free(msg);
  if (sig) free(sig);
  CAMLreturn(Val_bool(ok));
}

File: formal/fstar/ocaml-output/hacl_stubs.js
// hacl_stubs.js — js_of_ocaml / Node runtime realisation of the four VC
// crypto primitives that the NATIVE build realises in
// experimental_ocaml_glue/hacl_stubs.c over the vendored HACL* C.
//
// Crypto sourcing policy: skills/crypto-policy/SKILL.md — "Dont roll our own
// crypto!", "pursue HACL*".  These stubs contain NO cryptographic logic:
// every primitive (Ed25519 sign/verify/keygen, SHA-256/384) is a call into
// HACL*'s OWN official WebAssembly build (third_party/hacl-wasm/, the
// `hacl-wasm` npm artifacts), loaded by npm/factoidal/hacl-init.js.  The only
// logic here is the hex <-> byte marshalling around the OCaml string
// boundary — byte-for-byte the same contract as the native C stub:
//
//   caml_hacl_sha256(msg_bytes)                -> 64-char hex digest
//   caml_hacl_ed25519_secret_to_public(sk_hex) -> pk_hex (64 chars) | ""
//   caml_hacl_ed25519_sign(sk_hex, msg_hex)    -> sig_hex (128 chars) | ""
//   caml_hacl_ed25519_verify(pk_hex,msg_hex,sig_hex) -> bool (0/1)
//
// An empty-string return signals a malformed-length/odd-hex input (the F*
// wrapper VC.DataIntegrity treats that as failure); it NEVER signals a
// crypto verdict.  `verify` returns the HACL* verdict unchanged.
//
// This is a rule #11 assume-val realisation (a host-engine call-out to the
// SAME assume vals the native C realises), not new semantics.  It is loaded
// into the bundle by build-ocaml.sh's js_of_ocaml step, exactly like
// parquet_zstd_stubs.js.
//
// SECURITY: if the HACL* wasm module has not been initialised
// (globalThis.__factoidalHacl is unset), every primitive THROWS.  verify
// must never silently succeed when the crypto backend is missing — a
// verify that returns true without checking a signature is a security hole.
// Consumers MUST `await require('.../hacl-init.js').initHacl()` (Node) or
// call the browser initialiser before invoking any VC verify/sign path.

//Provides: caml_hacl_backend
function caml_hacl_backend() {
  var H = (typeof globalThis !== "undefined") ? globalThis.__factoidalHacl : undefined;
  if (!H) {
    throw new Error(
      "factoidal VC crypto: HACL* wasm backend not initialised. " +
      "Call `await require('./hacl-init.js').initHacl()` (Node) or the " +
      "browser initHacl() before any vcVerify/vcSign. Refusing to run " +
      "crypto without the verified backend (would be a security hole).");
  }
  return H;
}

// Normalise an OCaml-string argument that js_of_ocaml may hand us as a plain
// JS string (use-js-string=true) or as an MlBytes-like object, into a
// byte-string (each code unit is a byte value 0..255).  Matches the native
// stub's String_val: raw bytes, NOT a UTF-8 re-encoding.
//Provides: caml_hacl_bytestr
function caml_hacl_bytestr(s) {
  if (s && typeof s === "object") {
    if (typeof s.c === "string") return s.c;                 // MlBytes { c, l, t }
    if (typeof s.toString === "function") return s.toString();
  }
  return String(s);
}

// Byte-string -> Uint8Array (raw bytes; charCode & 0xff).
//Provides: caml_hacl_str_to_u8
//Requires: caml_hacl_bytestr
function caml_hacl_str_to_u8(s) {
  var bs = caml_hacl_bytestr(s);
  var n = bs.length;
  var out = new Uint8Array(n);
  for (var i = 0; i < n; i++) out[i] = bs.charCodeAt(i) & 0xff;
  return out;
}

// Hex OCaml string -> Uint8Array, or null on odd length / non-hex digit.
//Provides: caml_hacl_hex_to_u8
//Requires: caml_hacl_bytestr
function caml_hacl_hex_to_u8(hex) {
  var h = caml_hacl_bytestr(hex);
  if ((h.length % 2) !== 0) return null;
  var n = h.length / 2;
  var out = new Uint8Array(n);
  for (var i = 0; i < n; i++) {
    var hi = h.charCodeAt(2 * i);
    var lo = h.charCodeAt(2 * i + 1);
    var hv = hi >= 48 && hi <= 57 ? hi - 48
           : hi >= 97 && hi <= 102 ? hi - 87
           : hi >= 65 && hi <= 70 ? hi - 55 : -1;
    var lv = lo >= 48 && lo <= 57 ? lo - 48
           : lo >= 97 && lo <= 102 ? lo - 87
           : lo >= 65 && lo <= 70 ? lo - 55 : -1;
    if (hv < 0 || lv < 0) return null;
    out[i] = (hv << 4) | lv;
  }
  return out;
}

// Uint8Array -> lowercase hex string (no separators), matching the C stub.
//Provides: caml_hacl_u8_to_hex
function caml_hacl_u8_to_hex(bytes) {
  var hexd = "0123456789abcdef";
  var n = bytes.length;
  var out = new Array(n);
  for (var i = 0; i < n; i++) {
    var b = bytes[i] & 0xff;
    out[i] = hexd.charAt(b >> 4) + hexd.charAt(b & 0xf);
  }
  return out.join("");
}

// SHA-256 of the raw bytes of the OCaml string -> 64-char lowercase hex.
//Provides: caml_hacl_sha256
//Requires: caml_hacl_backend, caml_hacl_str_to_u8, caml_hacl_u8_to_hex
function caml_hacl_sha256(v_msg) {
  var H = caml_hacl_backend();
  var digest = H.SHA2.hash_256(caml_hacl_str_to_u8(v_msg))[0];
  return caml_hacl_u8_to_hex(digest);
}

// SHA-384 (exposed for parity with the SHA-2 family; VC uses SHA-256, but
// RDFC-1.0 / the SHA-2 assume vals want 384 too). 96-char lowercase hex.
//Provides: caml_hacl_sha384
//Requires: caml_hacl_backend, caml_hacl_str_to_u8, caml_hacl_u8_to_hex
function caml_hacl_sha384(v_msg) {
  var H = caml_hacl_backend();
  var digest = H.SHA2.hash_384(caml_hacl_str_to_u8(v_msg))[0];
  return caml_hacl_u8_to_hex(digest);
}

// Ed25519 secret key (32 bytes, hex) -> public key (32 bytes, hex), or ""
// on a malformed-length / non-hex secret key.
//Provides: caml_hacl_ed25519_secret_to_public
//Requires: caml_hacl_backend, caml_hacl_hex_to_u8, caml_hacl_u8_to_hex
function caml_hacl_ed25519_secret_to_public(v_sk_hex) {
  var H = caml_hacl_backend();
  var sk = caml_hacl_hex_to_u8(v_sk_hex);
  if (sk === null || sk.length !== 32) return "";
  var pk = H.Ed25519.secret_to_public(sk)[0];
  return caml_hacl_u8_to_hex(pk);
}

// Ed25519 sign: (sk_hex[32B], msg_hex) -> signature hex (64 bytes), or ""
// on a malformed secret key / non-hex input.
//Provides: caml_hacl_ed25519_sign
//Requires: caml_hacl_backend, caml_hacl_hex_to_u8, caml_hacl_u8_to_hex
function caml_hacl_ed25519_sign(v_sk_hex, v_msg_hex) {
  var H = caml_hacl_backend();
  var sk = caml_hacl_hex_to_u8(v_sk_hex);
  if (sk === null || sk.length !== 32) return "";
  var msg = caml_hacl_hex_to_u8(v_msg_hex);
  if (msg === null) return "";
  var sig = H.Ed25519.sign(sk, msg)[0];
  return caml_hacl_u8_to_hex(sig);
}

// Ed25519 verify: (pk_hex[32B], msg_hex, sig_hex[64B]) -> bool (0/1).
// A malformed-length / non-hex input is a verdict of false, never true.
//Provides: caml_hacl_ed25519_verify
//Requires: caml_hacl_backend, caml_hacl_hex_to_u8
function caml_hacl_ed25519_verify(v_pk_hex, v_msg_hex, v_sig_hex) {
  var H = caml_hacl_backend();
  var pk = caml_hacl_hex_to_u8(v_pk_hex);
  var msg = caml_hacl_hex_to_u8(v_msg_hex);
  var sig = caml_hacl_hex_to_u8(v_sig_hex);
  if (pk === null || pk.length !== 32) return 0;
  if (msg === null) return 0;
  if (sig === null || sig.length !== 64) return 0;
  var ok = H.Ed25519.verify(pk, msg, sig)[0];
  return ok ? 1 : 0;
}
